### PR TITLE
feat: add docker image releaser

### LINKED ISSUE #132: Generate a docker image for every release
Hi,

Is this possible that you maintain a docker image for the project ?
Every image that i find is clearly outdated.

### PR DESCRIPTION
**Pull request by anthonymq**
## What does this do / why do we need it?

It release a docker image for every release. I needed it in my Gitlab-ci to easily generate my changelog.

## How this PR fixes the problem?

It builds and push an image to Dockerhub.

## What should your reviewer look out for in this PR?

The dockerfile and the github actions files

## Check lists

* [ ] Test passed
* [ ] Coding style (indentation, etc)

## Additional Comments (if any)

You will need to create two secrets on the repo to authenticate to Dockerhub

## Which issue(s) does this PR fix?

fixes #132

### CODE REVIEW

**Comment by anthonymq** [comment]
Obviously we will have to rename the dockerhub artifact.
Gives https://hub.docker.com/layers/143400137/anthonymq/git-chglog/latest/images/sha256-69886af20e8c02f7ab7437f70cc1715251313ef3f5b0c10df824e65f66c06bca?context=explore

**Review by khos2ow — CHANGES_REQUESTED** [review]
Thank you @anthonymq, but we are using [GoReleaser](https://github.com/git-chglog/git-chglog/blob/master/.goreleaser.yml) to manage release process, and this should be incorporated into that process, and not a standalone one.

**Comment by anthonymq** [comment]
Yes sur, i put it in another file for testing purpose.
I'll change it.

**Comment by anthonymq** [comment]
Hi again, 
I've changed the github action to push on the github registry. 
Example : https://github.com/anthonymq?tab=packages&repo_name=git-chglog

**Comment by anthonymq** [comment]
> Thank you @anthonymq, but we are using [GoReleaser](https://github.com/git-chglog/git-chglog/blob/master/.goreleaser.yml) to manage release process, and this should be incorporated into that process, and not a standalone one.

Ok it's done.
You can verify on my repo https://github.com/anthonymq/git-chglog/runs/2228510121?check_suite_focus=true
There is an error because i can't upload the doc which is normal.
But the image is uploaded https://github.com/users/anthonymq/packages/container/package/git-chglog

There is an action that someone need to activate on the organization :

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -17,10 +17,16 @@ jobs:
...
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
       - name: Run GoReleaser
         uses: goreleaser/goreleaser-action@v2
         with:
           version: latest
           args: release --rm-dist
         env:
-          GITHUB_TOKEN: ${{ secrets.GORELEASER_TOKEN }}
```
Comment (khos2ow) [inline .github/workflows/release.yml:26]: Please revert back this line, `GORELEASER_TOKEN` should be used.

File: .goreleaser.yml [hunk 2]
```diff
@@ -47,4 +47,12 @@ brews:
     test: |
       system "#{bin}/git-chglog --help"
     install: |
-      bin.install "git-chglog"
\ No newline at end of file
+      bin.install "git-chglog"
+
+dockers:
```
Comment (khos2ow) [inline .goreleaser.yml:52]: Move `dockers`  before `brews`.

File: .goreleaser.yml [hunk 3]
```diff
@@ -47,4 +47,12 @@ brews:
...
+  - 
+    image_templates:
```
Comment (khos2ow) [inline .goreleaser.yml:54]: Make sure it's formatter properly:

```yaml
dockers:
  - image_templates:
      - "...."
    ...
```

File: .goreleaser.yml [hunk 4]
```diff
@@ -47,4 +47,12 @@ brews:
...
+    - "ghcr.io/git-chglog/git-chglog:{{ .Tag }}"
```
Comment (khos2ow) [inline .goreleaser.yml:55]: Replace `{{ .Tag }}` with `{{ .RawVersion }}`.

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,12 @@
+FROM docker.io/alpine:latest
```
Comment (khos2ow) [inline Dockerfile:1]: - `docker.io/` should be removed
- use `alpine:3.13.3` instead of `latest`

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+RUN apk add --no-cache git && \
+    mkdir /workdir
```
Comment (khos2ow) [inline Dockerfile:4]: There's no need for `/workdir`

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+COPY git-chglog /usr/local/bin/git-chglog
+
+WORKDIR /workdir
+RUN chmod +x /usr/local/bin/git-chglog
+
+ENTRYPOINT [ "/usr/local/bin/git-chglog" ]
+CMD [ "--help" ]
```
Comment (khos2ow) [inline Dockerfile:12]: Remove `CMD`, otherwise it's being forced to user.

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,12 @@
+FROM docker.io/alpine:latest
+
+RUN apk add --no-cache git && \
+    mkdir /workdir
+
+COPY git-chglog /usr/local/bin/git-chglog
+
+WORKDIR /workdir
+RUN chmod +x /usr/local/bin/git-chglog
```
Comment (khos2ow) [inline Dockerfile:9]: Technically speaking this line is not needed.

**Review by khos2ow — COMMENTED** [review]
Thank you @anthonymq, please see my comments below.

File: .github/workflows/release.yml (hunk 1 above)
Comment (anthonymq) [reply .github/workflows/release.yml:26]: We don't need a GORELEASER_TOKEN, GITHUB_TOKEN is injected automatically and has the rights.

File: Dockerfile (hunk 8 above)
Comment (anthonymq) [reply Dockerfile:9]: I tried without the 'chmod +x' and it didn't work.

File: .github/workflows/release.yml (hunk 1 above)
Comment (khos2ow) [reply .github/workflows/release.yml:26]: `GITHUB_TOKEN`, as you mentioned, is automatically injected but `secrets.GORELEASER_TOKEN` is a custom variable we are using for the release process here.

Comment (anthonymq) [reply .github/workflows/release.yml:26]: Is it still necessary ? GITHUB_TOKEN has the right to push packages on the repo and that's all we need.

Comment (khos2ow) [reply .github/workflows/release.yml:26]: Yes it is :)

Comment (anthonymq) [reply .github/workflows/release.yml:26]: Alright ;)

File: Dockerfile (hunk 6 above)
Comment (khos2ow) [reply Dockerfile]: On a second thought, I take back my original comment. It's still gonna be useful for users to have a `WORKDIR` set up. Please add it back as `/work`. And I apologize for it.

File: Dockerfile (hunk 8 above)
Comment (khos2ow) [reply Dockerfile:7]: If you let `goreleaser` build the image, the binary will be executable, but we can be on a safe side and leave it as is.

File: Dockerfile (hunk 6 above)
Comment (anthonymq) [reply Dockerfile:4]: Alright it is done 😉

**Comment by anthonymq** [comment]
example usage :  
``
docker run -v "$PWD":/workdir ghcr.io/anthonymq/git-chglog --tag-filter-pattern '^\d+\.\d+\.\d+$' "7.35.0"
 ``

File: .goreleaser.yml [hunk 9]
```diff
@@ -38,6 +38,14 @@ changelog:
 snapshot:
   name_template: "{{ .Tag }}-next"
 
+dockers:
+  - 
+    image_templates:
+      - "ghcr.io/git-chglog/git-chglog:{{ .RawVersion }}"
+      - "ghcr.io/git-chglog/git-chglog:latest"
+    ids:
+      - git-chglog
```
Comment (khos2ow) [inline .goreleaser.yml:47]:
```suggestion
dockers:
  - image_templates:
      - "ghcr.io/git-chglog/git-chglog:{{ .RawVersion }}"
      - "ghcr.io/git-chglog/git-chglog:latest"
```

**Review by khos2ow — COMMENTED** [review]
Thank you @anthonymq for the work so far. There's one more thing to implement, which to build and push `edge` tag on `master`. Please create `.github/workflows/docker.yml` and implement this in there. I should mention that this specific task has to be done outside of goreleaser process as it is bound to commits on master.

**Comment by khos2ow** [comment]
I had to make some adjustment to underlying machinery we needed to proceed with rest of this PR, specifically building `edge` image. I'm gonna post a review as soon as we finish #134. Thank you for the work so far @anthonymq and for your patience.

**Comment by anthonymq** [comment]
sorry.... I made a mistake with a wrong remote........... I'll open a new PR

**Pull request closed without merging**